# 예외 처리

## 예외 계층

### 기초

프로그램 실행 중 발생할 수 있는 예상치 못한 상황, 즉 예외(Exception)를 처리하기 위한 매커니즘을 제공. 이는 프로그램의 안정성과 신뢰성을 높이는데 중요한 역할을 함.

<img width="682" alt="image" src="https://github.com/user-attachments/assets/9e56bb77-4080-4eed-88ca-5fad5dbc38d7" />

- `Object`: 자바에서 기본형을 제외한 모든 것은 객체임. 모든 객체의 최상위 부모는 `Object` 이므로 예외의 최상위 부모도 `Object` 임.
- `Throwable`: 최상위 예외. 하위에 `Exception` 과 `Error` 가 있음.
- `Error`: 메모리 부족이나 심각한 시스템 오류와 같이 애플리케이션에서 복구가 불가능한 시스템 예외. 애플리케이션 개발자는 이 예외를 잡으려고 해서는 안된다.
- `Exception`: 체크 예외
  - 애플리케이션 로직에서 사용할 수 있는 실질적인 최상위 예외.
  - `Exception` 과 그 하위 예외는 모두 컴파일러가 체크하는 체크 예외이다. 단 `RuntimeException` 은 예외로 함.
- `RunTimeException`: 언체크 예외, 런타임 예외
  - 컴파일러가 체크 하지 않는 언체크 예외.
  - `RuntimeException` 과 그 자식 예외는 모두 언체크 예외이다.
  - `RuntimeException` 의 이름을 따라 `RuntimeException` 과 하위 언체크 예외를 **런타임 예외**라고 많이 부름.

> 체크 예외 vs 언체크 예외(런타임 예외)
>
> 체크 예외 외 발생한 예외를 개발자가 명시적으로 처리해야 함. 그렇지 않으면 컴파일 오류 발생. 언체크 예외는 개발자가 발생한 예외를 명시적으로 처리하지 않아도 된다.

※ 주의

상속 관계에서 부모 타입은 자식을 담을 수 있음. 예외 처리에도 적용되는데, 상위 예외를 `catch` 로 잡으면 하위 예외까지 함께 잡는다. 따라서 어플리케이션 로직에서는 `Throwable` 예외를 잡으면 안되는데, 앞서 이야기한 잡으면 안되는 `Error` 예외도 함께 잡을 수 있기 때문임. 어플리케이션 로직은 이런 이유로 `Exception` 부터 필요한 예외로 생각하고 잡으면 된다.

### 심화

<img width="627" alt="image" src="https://github.com/user-attachments/assets/333082c6-e34f-44dd-8f3d-66094c6c1078" />

예외를 단순히 오류 코드로 분류하는 것이 아니라, 예외를 계층화해 다양하게 만들면 더 세밀하게 예외를 처리할 수 있음.

- NetworkClientExceptionV3: NetworkClinet에서 발생하는 모든 예외는 이 예외의 자식이다.
- ConnectException: 연결 실패 시 발생하는 예외.
- SendException: 전송 실패 발생 예외.

이렇게 계층화 해두면 장점은

- 자바에서 예외는 객체이므로, 부모 예외를 잡거나 던지면 함께 자식 예외도 잡거나 던질 수 있음.
- 특정 예외를 잡아 처리하고 싶으면 하위 예외를 잡아 처리하면 된다.

<br/>

## 예외 기본 규칙

예외는 폭탄 돌리기와 같음. 예외가 발생하면 잡아서 처리하거나, 처리할 수 없으면 밖으로 던져야 함.

<img width="672" alt="image" src="https://github.com/user-attachments/assets/d8caa765-ce9a-430c-b95d-c242b58b2d1c" />

1. `Main` 은 `Service` 를 호출
2. Service는 Client 호출
3. Client에서 예외 발생
4. Clinet에서 예외를 처리하지 못하고 밖으로 던짐. 여기서 Client의 밖은 Client를 호출한 Service.
5. Service에 예외가 전달됨. Service에서 예외를 처리했다. 이후 어플리케이션 로직이 정상 흐름으로 동작.
6. 정상 흐름 반환

**2가지 기본 규칙**

1. 예외는 잡아서 처리하거나 밖으로 던져야 함.
2. 예외를 잡거나 던질 때 지정한 예외 뿐 아니라 그 예외의 자식들도 함께 처리할 수 있음.
   - `Exception` 을 `catch` 로 잡으면 그 하위 예외들도 모두 잡을 수 있음.
   - `Exception` 을 `throws` 로 잡으면 그 하위 예외들도 모두 잡을 수 있음.

**예외를 잡지못하고 계속 던지면?**

자바 `main()` 밖으로 예외를 던지면 예외 로그를 출력하면서 시스템이 종료된다.

<br/>

## 체크 예외

- `Exception` 과 그 하위 예외는 모두 컴파일러가 체크하는 체크 예외. 단, `RuntimeException` 은 예외로 한다.
- 체크 예외는 잡아서 처리하거나, 또는 밖으로 던지도록 선언해야 한다. 그렇지 않으면 컴파일 오류 발생.

```java
/**
 * Exception을 상속 받은 예외는 체크 예외가 된다.
 */
public class MyCheckedException extends Exception {
    public MyCheckedException(String message) {
        super(message);
    }
}
```

- 예외 클래스를 만들려면 예외를 상속 받으면 된다.
- `Exception` 을 상속 받은 예외는 체크 예외가 된다.

```java
public class Client {
    public void call() throws MyCheckedException {
        // 문제 상황
        throw new MyCheckedException("ex");
    }
}
```

- `throw 예외` 라고 하면 새로운 예외를 발생시킬 수 있다. 예외도 객체이기에 `new` 생성 후 예외를 발생시켜야 함.
- `throws 예외` 는 발생시킨 예외를 메서드 밖으로 던질 때 사용하는 키워드임.
- `throw`, `throws` 잘 구분해야 함.

```java
public class Service {
    Client client = new Client();

    /**
     * 예외를 잡아서 처리하는 코드
     */
    public void callCatch() {
        try {
            client.call();
        } catch (MyCheckedException e) {
            // 예외 처리 로직
            System.out.println("예외 처리, message= " + e.getMessage());
        }
        System.out.println("정상 흐름");
    }

    /**
     * 체크 예외를 밖으로 던지는 코드
     * 체크 예외는 예외를 잡지 않고 밖으로 던지려면 throws 예외를 메서드에 필수로 선언해야 함.
     */
    public void catchThrow() throws MyCheckedException {
        client.call();
    }
}
```

### 체크 예외의 장단점

예외를 잡아서 처리할 수 없을 때, 던지는 `throw 예외` 를 필수로 선언해야 함. 그렇지 않으면 컴파일 오류 발생.

- 장점: 개발자가 실수로 예외를 누락하지 않도록 컴파일러를 통해 문제를 잡아주는 훌륭한 안전 장치임. 이를 통해 개발자는 어떤 체크 예외가 발생하는 지 쉽게 파악 가능.
- 단점: 하지만 실제로 개발자가 모든 체크 예외를 반드시 잡거나 던지도록 해야하기 때문에 너무 번거로운 일이 됨. 크게 신경쓰고 싶지 않은 예외까지 모두 챙겨야 함.

<br/>

## 언체크 예외

말 그대로 컴파일러가 예외를 체크하지 않는다는 뜻. 체크 예외와 기본적으로 동일 하지만, 차이가 있다면 `throws` 를 선언하지 않고 생략할 수 있음. 자동으로 던진다.

```java
/**
 * RuntimeException을 상속받은 예외는 언체크 예외가 된다.
 */
public class MyUncheckedException extends RuntimeException {
    public MyUncheckedException(String message) {
        super(message);
    }
}

public class Client {
    public void call() {
        throw new MyUncheckedException("ex");
    }
}
```

```java
/**
 * Unchecked 예외는
 * 예외를 잡거나, 던지지 않아도 된다.
 * 예외를 잡지 않으면 자동으로 밖으로 던진다.
 */
public class Service {
    Client client = new Client();

    /**
     * 필요한 경우 예외를 잡아서 처리할 수 있음.
     */
    public void callCatch() {
        try {
            client.call();
        } catch (MyUncheckedException e) {
            // 예외 처리 로직
            System.out.println("예외 처리, message" + e.getMessage());
        }
        System.out.println("정상 로직");
    }
    /**
     * 예외를 잡지 않아도 된다. 자연스럽게 상위로 넘어간다.
     * 체크 예외와 다르게 throws 예외 선언을 하지 않아도 된다.
     */
    public void callThrow() {
        client.call();
    }
}
```

### 언체크 예외의 장단점

예외를 잡아서 처리할 수 없을 때, 던지는 `throw 예외` 를 생략.

- 장점: 신경쓰고 싶지 않은 언체크 예외를 무시 가능. 체크 예외의 경우 처리할 수 없는 예외를 밖으로 던지려면 항상 `throws 예외` 를 선언해야 하지만, 언체크는 이 부분 생략 가능.
- 단점: 개발자 실수로 예외 누락 할 수 있음. 반면 체크 예외는 컴파일러를 통해 예외 누락을 잡아줌.

> ⭐️ 현대는 체크 예외를 거의 쓰지 않음.
>
> 예전엔 체크 예외를 쓰는게 맞았음. 컴파일러가 걸러주기 때문. 하지만, 요즘 IDE가 잘 되어있기 때문임.

<br/>

## 실무 예외 처리 방안

**처리할 수 없는 예외**

상대 네트워크 서버에 문제가 발생하면 어플리케이션 연결 오류, DB 접속 실패와 같은 예외 발생. 이러면 예외를 해결할 수 없음. 예외를 잡아 다시 호출을 시도해도 같은 오류 반복 된다.

고객에게는 "현재 시스템에 문제가 있습니다." 라는 오류 메시지 보여줌. 웹이라면 오류 페이지 보여주면 된다. 그리고 내부 개발자가 문제 상황을 빠르게 인지할 수 있도록, 오류에 대한 로그를 남겨두어야 함.

**체크 예외의 부담**

체크 예외는 개발자가 실수로 놓칠 수 있는 예외를 컴파일러 체크. 하지만 예외의 종류가 많아지고, 프로그램이 복잡해지면서 체크 예외를 사용하는 것이 점점 더 부담임.

### 체크 예외 사용 시나리오

<img width="624" alt="image" src="https://github.com/user-attachments/assets/2db71c7c-d8f5-46ca-b73a-f97c92fad38e" />

- 실무에서는 수 많은 라이브러리 사용, 또 다양한 외부 시스템과 연동.
- 사용하는 각각의 클래스들이 자신만의 예외를 모두 체크 예외로 만들어 전달한다고 가정하자.

<img width="628" alt="image" src="https://github.com/user-attachments/assets/b2d021f6-2566-478f-85ed-788239151991" />

- 이 경우 `Service` 는 호출하는 곳에서 던지는 체크 예외들을 처리해야 함. 만약 처리할 수 없다면 밖으로 던저야 함.

모든 체크 예외를 잡아 처리해야함. 하지만, 본인이 처리할 수 없는 예외 들이라면 밖으로 던지는 것이 더 나은 결정임.

```java
class Service {
  void sendMessage(String data) throws NetworkException, DatabaseException, ... { ... }
}
```

- 이렇게 모든 체크 예외를 하나씩 다 밖으로 던져야 함.
- 라이브러리가 늘어날 수록 다루어야 하는 예외도 더 많아짐. 개발자 입장에서는 번거로움.

문제는 끝이 아님. 만약 `Facade` 라는 클래스가 있다고 가정해보자.

<img width="621" alt="image" src="https://github.com/user-attachments/assets/4ea2dc6f-8931-4825-b098-42ce9b16b0f4" />

- 이 경우 `Facade` 클래스에서도 이런 예외를 복구할 수 없음. `Facade` 클래스도 밖으로 던져야 함.
- 결국 중간에 모든 클래스에서 예외를 계속 밖으로 던지는 지저분한 코드가 만들어짐.
- `throws` 로 발견한 모든 예외를 다 밖으로 던지는 것.

그러면 `Exception` 을 던져버리면 수 많은 Exception들을 한번에 잡을 수 있지 않을까? -> 안된다.

**throws Exception의 문제**

최상위 타입이므로 모든 체크 예외를 다 밖으로 던지는 문제 발생. 결과적으로 체크 예외의 최상위 타입인 `Exception` 을 던지게 되면 다른 체크 예외를 체크할 수 있는 기능이 무효화, 중요한 체크 예외를 다 놓치게 된다.

### 언체크(런타임) 예외 사용 시나리오

<img width="625" alt="image" src="https://github.com/user-attachments/assets/330bfdf1-1c9f-4335-bfa4-c707136e5f05" />

- 이번에는 `Service` 에서 호출하는 클래스들이 언체크(런타임) 예외를 던진다고 가정
- `NetworkException`, `DatabaseException` 은 잡아도 복구할 수 없음. 언체크 예외이므로 이런 경우 무시하면 된다.

즉, 언체크 예외를 공통에서 잡기로 하고 일부 언체크 예외를 잡아 처리하는 것으로 한다.

```java
public static void main(String[] args) {
    NetworkService networkService = new NetworkService();

    Scanner scanner = new Scanner(System.in);
    while (true) {
        System.out.println("전송할 문자: ");
        String input = scanner.nextLine();
        if (input.equals("exit")) {
            break;
        }

        try {
            networkService.sendMessage(input);
        } catch (Exception e) {
            exceptionHandler(e);
        }

        System.out.println();
    }
    System.out.println("프로그램을 정상 종료합니다.");
}

// 공통 예외 처리
private static void exceptionHandler(Exception e) {
    System.out.println("사용자 메시지: 죄송합니다. 알 수 없는 문제가 발생했습니다.");
    System.out.println("== 개발자용 디버깅 메시지 ==");
    e.printStackTrace(System.out); // 스택 트레이스 출력
//        e.printStackTrace();

    // 필요하면 예외 별로 별도 추가 처리 가능
    if(e instanceof SendException sendEx) {
        System.out.println("[전송 오류] 전송 데이터: " + sendEx.getSendData());
    }
}
```

### exceptionHandler()

- 해결할 수 없는 예외가 발생하면 사용자에게는 시스템 내 알 수 없는 문제가 발생했다고 알리는 것이 좋음
  - 사용자가 디테일한 오류 코드나 오류 상황까지 모두 이해할 필요 없음.
- 개발자가 빨리 문제를 찾고 디버깅할 수 있도록 오류 메시지를 남겨두어야 함.
- 예외도 객체이므로 `instanceof` 와 같이 예외 객체 타입 확인 후 별도의 추가 처리 가능

### e.printStackTrace()

- 예외 메시지와 스택 트레이스를 출력할 수 있음.
- 예외가 발생 지점을 역으로 추적 가능
- `e.printStackTrace(System.out)` 을 사용해 표준 출력으로 내보냄.
- `e.printStackTrace()` 를 사용하면 `System.err` 이라는 표준 오류에 결과를 출력함.
  - IDE에서는 `System.err` 로 출력하면 출력 결과를 빨간색으로 보여줌.
  - 일반적으로 이 방법을 사용함.

> `System.out`, `System.err` 둘다 콘솔에 출력되지만 서로 다른 흐름을 통해 출력 됨. 둘을 함께 사용하면 출력 순서를 보장하지 않음.
>
> 실무에서는 Slf4J, logback 같은 별도 로그 라이브러리를 사용함. 그런데 `e.printStatckTrace()` 를 직접 호출하면 결과가 콘솔에만 출력된다. 이렇게 되면 서버에서 로그 확인이 어려움. 서버에서는 파일로 로그를 확인해야 함. 콘솔에 바로 결과를 출력하는 `e.printStatckTrace()` 를 잘 사용하지 않음.

<br/>

## Try-with-resource

어플리케이션에서 외부 자원을 사용하는 경우 반드시 외부 자원을 해제 해야 함. 따라서 `finally` 구문을 반드시 사용해야 함.

하지만 이패턴이 반복되면서 자바 7에서는 Try with resources 편의 기능을 도압함. 이름 그대로 try에서 자원을 함께 사용한다는 뜻임. 자원은 try가 끝나면 반드시 종료 해서 반납해야 하는 외부 자원을 뜻함.

`AutoCloseable` 인터페이스를 구현해야 함.

```java
package java.lang;
public interface AutoCloseable {
   void close() throws Exception;
}
```

이 인터페이스를 구현하면 Try with resources를 사용할 때 `try` 가 끝나는 시점에 `close()` 가 자동으로 호출된다. 그리고 다음과 같이 Try with resources 구문을 사용하면 된다.

```java
try (Resource resource = new Resource()) {
  // 리소스를 사용하는 코드
}
```

**장점**

- **리소스 누수 방지**: 모든 리소스가 제대로 닫히도록 보장한다. 실수로 `finally` 블록을 적지 않거나, `finally` 블럭 안에서 자원 해제 코드를 누락하는 문제들을 예방할 수 있다.
- **코드 간결성 및 가독성 향상**: 명시적인 `close()` 호출이 필요 없어 코드가 더 간결하고 읽기 쉬워진다.
- **스코프 범위 한정**: 예를 들어 리소스로 사용되는 `client` 변수의 스코프가 `try` 블럭 안으로 한정된다. 따라서 코드 유지보수가 더 쉬워진다.
- **조금 더 빠른 자원 해제**: 기존에는 try catch finally로 catch 이후에 자원을 반납했다. Try with resources 구분은 `try` 블럭이 끝나면 즉시 `close()` 를 호출한다.
